### PR TITLE
[TH2-1988] Split connection into publishing and consuming

### PR DESCRIPTION
**Pull request by Sergey-Sitnikov-Exactpro**

### CODE REVIEW

**Review by OptimumOpium — CHANGES_REQUESTED** [review]
The current problem is that the exception handler and connection recovery handler uses the same atomic counter. So, the triggering will work incorrectly. We should separate the connections for publication and subscription here as well. So, please set common settings for ConnectionFactory and then use specific listeners before creating a connection

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java [hunk 1]
```diff
@@ -62,31 +63,30 @@
...
-            connectionRecoveryAttempts.set(0);
-            metrics.getReadinessMonitor().enable();
-            LOGGER.debug("Set RabbitMQ readiness to true");
-        }
-
-        @Override
-        public void handleRecoveryStarted(Recoverable recoverable) {}
-    };
+    private final RecoveryListener publishRecoveryListener = getRecoveryListener(PUBLISH_CONNECTION_NAME);
+    private final RecoveryListener consumehRecoveryListener = getRecoveryListener(CONSUME_CONNECTION_NAME);
```
Comment (andreydrobynin) [inline src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java:89]: Type in consume**h**RecoveryListener.
And maybe we don't need variables `publishRecoveryListener` and `consumehRecoveryListener` and we can use inline in `.addRecoveryListener()`?

Comment (andreydrobynin) [inline src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java:89]: Type in consume**h**RecoveryListener.
And I think we don't need fields `publishRecoveryListener` and `consumehRecoveryListener` and we can use inline in `addRecoveryListener()`?

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java [hunk 2]
```diff
@@ -377,6 +258,220 @@ private void basicAck(Channel channel, long deliveryTag) throws IOException {
...
+                tmpMetrics = publishMetrics;
+                factory.setSharedExecutor(publishSharedExecutor);
+                break;
+            case CONSUME_CONNECTION_NAME:
+                tmpMetrics = consumeMetrics;
+                factory.setSharedExecutor(consumeSharedExecutor);
+                break;
+            default:
+                throw new IllegalStateException("Unexpected value: " + connectionName);
+        }
```
Comment (andreydrobynin) [inline src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java:336]: Extract lines 324-336 to method `getHealthMetrics()` like `getRecoveryListener()`?

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java [hunk 3]
```diff
@@ -377,6 +258,220 @@ private void basicAck(Channel channel, long deliveryTag) throws IOException {
...
+        Connection tmpConnection;
```
Comment (andreydrobynin) [inline src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java:337]: Rename to `connection`?

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java [hunk 4]
```diff
@@ -377,6 +258,220 @@ private void basicAck(Channel channel, long deliveryTag) throws IOException {
...
+            throw new IllegalStateException("Cannot recover connection to RabbitMQ");
+        }
+        onFailedRecoveryConnection.run();
+        return false;
+    }
+
+    private int getRecoveryDelay(String connectionName) {
+        int tmpCountTriesToRecovery;
+        switch (connectionName) {
+        case PUBLISH_CONNECTION_NAME:
```
Comment (andreydrobynin) [inline src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java:442]: Format
```
switch (connectionName) {
    case PUBLISH_CONNECTION_NAME:
...
```

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java [hunk 5]
```diff
@@ -377,6 +258,220 @@ private void basicAck(Channel channel, long deliveryTag) throws IOException {
...
+                turnOffReadiness(exception);
+            }
+
+            @Override
+            public void handleTopologyRecoveryException(Connection conn, Channel ch, TopologyRecoveryException exception) {
+                turnOffReadiness(exception);
+            }
+
+            private void turnOffReadiness(Throwable exception){
+                switch (connectionName) {
```
Comment (andreydrobynin) [inline src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java:304]: I think there are too many `switch (connectionName)`.
Maybe it's better to add `PublishConnectionManager` and `ConsumeConnectionManager` with `ConnectionManager` as parent (or without parent), encaplulate all logic in them and create them in `ConnectionManager` constructor?

Comment (andreydrobynin) [reply src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java]: Sorry that I confused you, looks like now `PublishConnectionManager` and `ConsumeConnectionManager` are very similar.
I meant that we will not duplicate code in many `switch (connectionName)` and also in these "managers".

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java (hunk 2 above)
Comment (andreydrobynin) [reply src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java]: not actual

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java (hunk 3 above)
Comment (andreydrobynin) [reply src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java]: not actual

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java (hunk 4 above)
Comment (andreydrobynin) [reply src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java]: not actual

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java (hunk 5 above)
Comment (andreydrobynin) [reply src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java]: not actual

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java [hunk 6]
```diff
@@ -58,35 +58,27 @@
...
     private final String subscriberName;
     private final AtomicInteger nextSubscriberId = new AtomicInteger(1);
-    private final ExecutorService sharedExecutor = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
-            .setNameFormat("rabbitmq-shared-pool-%d")
+    private final ExecutorService publishSharedExecutor = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+            .setNameFormat("rabbitmq-" + PUBLISH_CONNECTION_NAME + "-shared-pool-%d")
+            .build());
+    private final ExecutorService consumeSharedExecutor = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+            .setNameFormat("rabbitmq-" + CONSUME_CONNECTION_NAME + "-shared-pool-%d")
             .build());
```
[the remark this answers is not in the document]
Comment (andreydrobynin) [reply src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java]: done

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Pull request merged**